riscv/src/register/macros.rs

```diff
@@ -449,6 +449,34 @@ macro_rules! read_composite_csr {
     };
 }
 
+/// Convenience macro to write a composite value to a CSR register.
+///
+/// - `RV32`: writes 32-bits into `hi` and 32-bits into `lo` to create a 64-bit value
+/// - `RV64`: writes a 64-bit value into `lo`
+#[macro_export]
+macro_rules! write_composite_csr {
+    ($hi:expr, $lo:expr) => {
+        /// Writes the CSR as a 64-bit value
+        #[inline]
+        pub unsafe fn write64(bits: u64) {
+            match () {
+                #[cfg(target_arch = "riscv32")]
+                () => {
+                    // Setting low value to max first prevents the register from being temporarily
+                    // reduced by a transaction that is intended for increasing the total
+                    // value.
+                    $lo(usize::MAX);
+                    $hi((bits >> 32) as usize);
+                    $lo(bits as usize);
```

Comment:
So, there's a little bit of a trick here that I took from a RISC-V specification related to `mtimer`. For `mtimer` this trick is very necessary to avoid mtimer driven compare matches that cause interrupts (e.g., any CLINT / base spec interrupt architecture).

I don't think mcycle and minstret natively drive any compare matches in a sensible implementation but it may or may not be a good idea to do it like this just be safe. What do you think?

Another valid--possibly better--interpretation would be to say that this macro just sets the registers and if you want to have specific guarantees you write the code manually.

Reply:
Technically, I guess it's possible you could want to have the opposite guarantee of "register should not temporarily increase for a transaction that is intended for decreasing the total value". I'm therefore predisposed against doing it like shown and instead just doing the writes naively.

Reply:
I don't think implicitly writing a value to a CSR is a good idea. In this specific case, it may be safe, however, it could have different semantics depending on the CSR.

As an example, `mstatus` uses `1` values to enable interrupts. So, a user of `write64` would temporarily enable interrupts, even if explicitly writing values to disable interrupts. If an interrupt happens to fire in the short time between the two `$lo` writes, the system could be put into an inoperable state.

If `write_composite_csr` is intended to be general-use, I would suggest removing the implicit first write.